client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/privileges/IndicesPrivileges.java

```diff
@@ -0,0 +1,324 @@
+     * all the document's fields are granted access to. Can also be empty, in which
+     * case no fields are granted access to.
+     */
+    public @Nullable Set<String> getGrantedFields() {
+        return this.grantedFields;
+    }
+
+    /**
+     * The document fields that cannot be accessed or queried. Can be null or empty,
+     * in which case no fields are denied.
+     */
+    public @Nullable Set<String> getDeniedFields() {
+        return this.deniedFields;
+    }
+
+    /**
+     * A query limiting the visible documents in the indices. Can be null, in which
+     * case all documents are visible.
+     */
+    public @Nullable String getQuery() {
+        return this.query;
+    }
+
+    /**
+     * If {@code true} some documents might not be visible. Only the documents
+     * matching {@code query} will be readable.
+     */
+    public boolean isUsingDocumentLevelSecurity() {
+        return query != null;
+    }
+
+    /**
+     * If {@code true} some document fields might not be visible.
+     */
+    public boolean isUsingFieldLevelSecurity() {
+        return limitsGrantedFields() || hasDeniedFields();
+    }
+
+    private boolean hasDeniedFields() {
+        return deniedFields != null && false == deniedFields.isEmpty();
+    }
+
+    private boolean limitsGrantedFields() {
+        // we treat just '*' as no FLS since that's what the UI defaults to
+        if (grantedFields == null || (grantedFields.size() == 1 && grantedFields.iterator().next().equals("*"))) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        IndicesPrivileges that = (IndicesPrivileges) o;
+        return indices.equals(that.indices)
+                && privileges.equals(that.privileges)
+                && Objects.equals(grantedFields, that.grantedFields)
+                && Objects.equals(deniedFields, that.deniedFields)
+                && Objects.equals(query, that.query);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(indices, privileges, grantedFields, deniedFields, query);
+    }
+
+    @Override
+    public String toString() {
+        try {
+            return XContentHelper.toXContent(this, XContentType.JSON, true).utf8ToString();
+        } catch (IOException e) {
+            throw new RuntimeException("Unexpected", e);
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(NAMES.getPreferredName(), indices);
+        builder.field(PRIVILEGES.getPreferredName(), privileges);
+        if (isUsingFieldLevelSecurity()) {
+            builder.startObject(FIELD_PERMISSIONS.getPreferredName());
+            if (limitsGrantedFields()) {
```

Comment:
I don't think this `if` is correct.
If granted is `*`, then we won't output anything, but the server requires a "grant" if there is an "except".

So this will fail if you try to put the equivalent of
```
"field_security": { 
   "grant": ["*"] , 
   "except": [ "secret.*" ]
}
```

Reply:
good catch!